### PR TITLE
Add ServiceDiscoveryRole configuration to toggle service discovery via EndpointSlice vs Endpoints

### PR DESCRIPTION
**Pull request by thaniri**
## Description

This is a solution to https://github.com/prometheus-operator/prometheus-operator/issues/3862 and is a replacement for the proposed solution https://github.com/prometheus-operator/prometheus-operator/pull/6663. This builds from the work @haanhvu started here https://github.com/prometheus-operator/prometheus-operator/pull/6518.

This PR introduces a `serviceDiscoveryRole` field to the specs for `Prometheus` and `PrometheusAgent` via the `CommonPrometheusFields` configuration pattern.

This will necessarily require a patch to [kube-prometheus-stack](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack) that looks something like this:

<details>
  <summary>kube-prometheus-stack patch</summary>
  
  Shows what changes were required to the kube-prometheus-stack helm chart to make use of the `serviceDiscoverRole` configuration.
  
  ```
  diff --git a/charts/kube-prometheus-stack/charts/crds/crds/crd-prometheusagents.yaml b/charts/kube-prometheus-stack/charts/crds/crds/crd-prometheusagents.yaml
  index caa24793..e332c5d4 100644
  --- a/charts/kube-prometheus-stack/charts/crds/crds/crd-prometheusagents.yaml
  +++ b/charts/kube-prometheus-stack/charts/crds/crds/crd-prometheusagents.yaml
  @@ -5535,6 +5535,15 @@ spec:
                     Only valid in Prometheus versions 2.45.0 and newer.
                   format: int64
                   type: integer
  +              serviceDiscoveryRole:
  +                description: |-
  +                  Defines the service discovery mechanism used to discover scrape targets.
  +                  If set, the value should be either "Endpoints" or "EndpointSlice".
  +                default: Endpoints
  +                enum:
  +                - Endpoints
  +                - EndpointSlice
  +                type: string
                 scrapeClasses:
                   description: |-
                     List of scrape classes to expose to scraping objects such as
  diff --git a/charts/kube-prometheus-stack/charts/crds/crds/crd-prometheuses.yaml b/charts/kube-prometheus-stack/charts/crds/crds/crd-prometheuses.yaml
  index 2dc5ba9c..c511e988 100644
  --- a/charts/kube-prometheus-stack/charts/crds/crds/crd-prometheuses.yaml
  +++ b/charts/kube-prometheus-stack/charts/crds/crds/crd-prometheuses.yaml
  @@ -6839,6 +6839,15 @@ spec:
                     Only valid in Prometheus versions 2.45.0 and newer.
                   format: int64
                   type: integer
  +              serviceDiscoveryRole:
  +                description: |-
  +                  Defines the service discovery mechanism used to discover scrape targets.
  +                  If set, the value should be either "Endpoints" or "EndpointSlice".
  +                default: Endpoints
  +                enum:
  +                - Endpoints
  +                - EndpointSlice
  +                type: string
                 scrapeClasses:
                   description: |-
                     List of scrape classes to expose to scraping objects such as
  diff --git a/charts/kube-prometheus-stack/templates/prometheus/prometheus.yaml b/charts/kube-prometheus-stack/templates/prometheus/prometheus.yaml
  index ae95fd02..400dce9f 100644
  --- a/charts/kube-prometheus-stack/templates/prometheus/prometheus.yaml
  +++ b/charts/kube-prometheus-stack/templates/prometheus/prometheus.yaml
  @@ -185,6 +185,9 @@ spec:
   {{ else }}
     podMonitorNamespaceSelector: {}
   {{- end }}
  +{{- if .Values.prometheus.prometheusSpec.serviceDiscoveryRole }}
  +  serviceDiscoveryRole: {{ .Values.prometheus.prometheusSpec.serviceDiscoveryRole }}
  +{{- end }}
   {{- if .Values.prometheus.prometheusSpec.probeSelector }}
     probeSelector:
   {{ tpl (toYaml .Values.prometheus.prometheusSpec.probeSelector | indent 4) . }}
  diff --git a/charts/kube-prometheus-stack/values.yaml b/charts/kube-prometheus-stack/values.yaml
  index 5dd92fbd..75c03cdf 100644
  --- a/charts/kube-prometheus-stack/values.yaml
  +++ b/charts/kube-prometheus-stack/values.yaml
  @@ -3365,6 +3365,7 @@ prometheus:
     ## ref: https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#prometheusspec
     ##
     prometheusSpec:
  +    serviceDiscoveryRole: Endpoints
       ## Statefulset's persistent volume claim retention policy
       ## pvcDeleteOnStsDelete and pvcDeleteOnStsScale determine whether
       ## statefulset's PVCs are deleted (true) or retained (false) on scaling down
  ```
</details>

## Type of change

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [x] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

This can be tested by deploying the `kube-prometheus-stack` helm chart with the above patch. When `serviceDiscoveryRole` is set to `Endpoints` then the generated scrape configs will have a `kubernetes_sd_role` of `endpoints`. Otherwise if `serviceDiscoveryRole` is set to `EndpointSlice` then the generated scrape configs will have a `kubernetes_sd_role` of `endpointslices`. If `serviceDiscoveryRole` is unset, or is set to an incorrect value, then the generated scrape configs will have a `kubernetes_sd_role` of `endpoints`.

## Changelog entry

```release-note
[ENHANCEMENT] Add ServiceDiscoveryRole configuration for use of Endpoints vs EndpointSlice.
```

### CODE REVIEW

**Comment by ArthurSens** [comment]
Just tagging @haanhvu here! Once you have the time, could you review this PR? :)

File: pkg/prometheus/promcfg_test.go
Comment (haanhvu) [inline pkg/prometheus/promcfg_test.go:914]: This check is valid, so we can keep it. However, it doesn't actually test the change introduced in this PR.

Looking at the implementation in this PR, the change happens at the `sync()` function of Prometheus/PrometheusAgent operator. Basically, `sync()` checks the value of `ServiceDiscoveryRole`, then set `endpointSliceSupported` accordingly. Meanwhile, this check is actually calling `generateK8SSDConfig()`, which is not a change in this PR.

So to check the change of this PR, could you add a new test to verify that `sync()` sets `endpointSliceSupported` according to the value of `ServiceDiscoveryRole`? Thankfully we already had [TestServiceMonitorWithEndpointSliceEnable](https://github.com/prometheus-operator/prometheus-operator/blob/ad006f4d607c99062d66936a8c02dd8bbaa6b40e/pkg/prometheus/promcfg_test.go#L1535C6-L1535C47) that verifies EndpointSlice is correctly used if `endpointSliceSupported == true`, so we don't need to care about this later step.

**Review by haanhvu — COMMENTED** [review]
Thanks for helping it move forward! I left a few comments. Feel free to ask for more information if you need.

File: pkg/prometheus/agent/operator.go
Comment (thaniri) [reply pkg/prometheus/agent/operator.go]: - I have updated the log level to Warn
- I have added additional information for the end user to know what to configure if that warning message appears
- Not technically required as Golang will default bools that are not explicitly given a value to false. But it is a good catch for the inconsistency and I think being explicit is a good practice so I have added this suggestion.

Please see 70d258855 for my changes for these comments.

Comment (haanhvu) [reply pkg/prometheus/agent/operator.go]: > Not technically required as Golang will default bools that are not explicitly given a value to false. But it is a good catch for the inconsistency and I think being explicit is a good practice so I have added this suggestion.

Yes the default bool is false. But in this case `endpointSliceSupported` is set through 2 steps:

- Step 1: Check if the version supports EndpointSlice: https://github.com/prometheus-operator/prometheus-operator/blob/ad006f4d607c99062d66936a8c02dd8bbaa6b40e/pkg/prometheus/agent/operator.go#L320-L324
- Step 2: Check the value of `ServiceDiscoveryRole` (this PR's change)

So there may be cases when `endpointSliceSupported` is true in step 1, but false in step 2. E.g., if the version supports EndpointSlice but users still need to use Endpoints. So we need to update `endpointSliceSupported` in step 2 in any cases.

The reason why step 1 is not enough and we need step 2 is explained in https://github.com/prometheus-operator/prometheus-operator/issues/4634

File: pkg/apis/monitoring/v1/prometheus_types.go
Comment (thaniri) [reply pkg/apis/monitoring/v1/prometheus_types.go]: I've seen it used as a pattern for optional elements for JSON objects to be represented as pointers to values in Golang. As this is an easy change and makes sense, I have implemented this suggestion in [2ddd628](https://github.com/prometheus-operator/prometheus-operator/pull/6672/commits/2ddd628152e07af43f0d34dd241fbedbc84c22df)

File: pkg/prometheus/server/operator.go
Comment (thaniri) [reply pkg/prometheus/server/operator.go]: Please see comment in https://github.com/prometheus-operator/prometheus-operator/pull/6672#discussion_r1652041060

Suggestion has been accepted.

File: pkg/prometheus/testdata/K8SSDConfigGenerationThree.golden
Comment (thaniri) [reply pkg/prometheus/testdata/K8SSDConfigGenerationThree.golden:5]: I manually created this file. After running `make test-unit-update-golden` nothing changes.

File: pkg/prometheus/promcfg_test.go
Comment (thaniri) [reply pkg/prometheus/promcfg_test.go]: Thank you for pointing this out. I have removed this setting in [5cab2fc](https://github.com/prometheus-operator/prometheus-operator/pull/6672/commits/5cab2fc5481a9816f3adfeb919c8353435322c55)

I have opted not to add `assert.False(t, cg.endpointSliceSupported)` as it will default to false.

File: pkg/prometheus/promcfg_test.go
Comment (thaniri) [reply pkg/prometheus/promcfg_test.go:913]: Hello @haanhvu , thank you for pointing out that an existing function `TestServiceMonitorWithEndpointSliceEnable` exists which tests that a scrape configuration is properly generated if EndpointSlices are supported.

I am having trouble finding where in the code we have an example of the `sync()` function being tested. As the function is quite large it would probably require an integration test or extensive mocking. You are correct that the core logic of this PR is in the reconciliation loop evaluating if both 1) the Kubernetes API supports EndpointSlices and 2) the user has configured their prometheus-operator to use EndpointSlices for service discovery. **Would you be able to push me in the correct direction to test the reconciliation loop of Prometheus?**

The key lines I can identify are:

1. Where `sync()` is setting the `endpointSliceSupported` field on the `Operator` object: https://github.com/thaniri/prometheus-operator/blob/1da76da45e78597f973ba1341780c2e406786fd2/pkg/prometheus/server/operator.go#L813-L825
2. The function that creates new `Operator` objects: https://github.com/thaniri/prometheus-operator/blob/1da76da45e78597f973ba1341780c2e406786fd2/pkg/prometheus/server/operator.go#L123
3. Where the `New()` function is called from: https://github.com/thaniri/prometheus-operator/blob/1da76da45e78597f973ba1341780c2e406786fd2/cmd/operator/main.go#L352
4. Where the `Operator` object actually starts https://github.com/thaniri/prometheus-operator/blob/1da76da45e78597f973ba1341780c2e406786fd2/cmd/operator/main.go#L531-L533

I am envisioning instantiating an `Operator` and calling the `sync` function on it to see if the `endpointSliceSupported` field on the Operator actually gets updated.

**If there is no current testing on the `sync()` function, could that perhaps be its own issue?**

File: pkg/prometheus/agent/operator.go
Comment (thaniri) [reply pkg/prometheus/agent/operator.go]: Thank you for the further explanation. This is a good thing to catch and ensure predictable behaviour out of the operator.

Please let me know if there is further work you would like to see for this particular suggestion.

File: pkg/prometheus/promcfg_test.go
Comment (haanhvu) [reply pkg/prometheus/promcfg_test.go:913]: Let me think how we can test this change as simply as we can. @simonpasquier @ArthurSens any ideas?

**Comment by haanhvu** [comment]
@thaniri I'll try to take the next review this week. In the meantime could you rebase?

**Review by simonpasquier — COMMENTED** [review]
This is a great start!
IMHO the decision about which discovery role to use should be delegated to `ConfigGenerator`. It has all the information needed to make the decision: it has the Prometheus/PrometheusAgent spec (Prom version + discovery role) + the information about whether endpointslice is supported.

Also the `endpointSliceSupported` field in the struct is badly named, it should rather be `useEndpointSlice` (or something simliar) and initialized in NewConfigGenerator() based on the inputs:
https://github.com/prometheus-operator/prometheus-operator/blob/ad006f4d607c99062d66936a8c02dd8bbaa6b40e/pkg/prometheus/promcfg.go#L60-L68

File: pkg/apis/monitoring/v1/prometheus_types.go
Comment (thaniri) [reply pkg/apis/monitoring/v1/prometheus_types.go]: This suggestion has been added. Thank you for your suggestion.

File: pkg/prometheus/server/operator.go
Comment (thaniri) [reply pkg/prometheus/server/operator.go:813]: I have separately made `cpf.ServiceDiscoveryRole` a pointer to a string instead of a string so it might not be necessary to use the `ptr` package to check the service discovery role in this switch statement.

File: pkg/apis/monitoring/v1/prometheus_types.go
Comment (thaniri) [reply pkg/apis/monitoring/v1/prometheus_types.go:1]: Thanks for pointing out this pattern. I have implemented this in [ab692b0](https://github.com/prometheus-operator/prometheus-operator/pull/6672/commits/ab692b0d9d94cd0d011a9b20a8497dc701aca870)

File: pkg/prometheus/server/operator.go
Comment (thaniri) [reply pkg/prometheus/server/operator.go]: This improves the code clarity. I have modified this in [3c0a27a](https://github.com/prometheus-operator/prometheus-operator/pull/6672/commits/3c0a27a93a4e2a7583e5e5d744da68c8d0fe9ec8)

File: pkg/prometheus/promcfg_test.go
Comment (thaniri) [reply pkg/prometheus/promcfg_test.go:913]: @simonpasquier recommended moving out the core logic of this PR into `NewConfigGenerator()` in [this comment](https://github.com/prometheus-operator/prometheus-operator/pull/6672#pullrequestreview-2130701946). This is a very good suggestion as we will be able to easily unit test this.

I have modified TestServiceMonitorWithEndpointSliceEnable to account for this logic in [8ca42f5](https://github.com/prometheus-operator/prometheus-operator/pull/6672/commits/8ca42f506ccd1d3ae786aee76c004c9c65dee7d0)

File: pkg/apis/monitoring/v1/prometheus_types.go
Comment (ArthurSens) [inline pkg/apis/monitoring/v1/prometheus_types.go:779]:
```suggestion
	ServiceDiscoveryRole *ServiceDiscoveryRole `json:"serviceDiscoveryRole,omitempty"`
```

Let's use the newly created Enum :)

File: pkg/prometheus/promcfg.go
Comment (ArthurSens) [inline pkg/prometheus/promcfg.go:99]:
```suggestion
	case monitoringv1.EndpointSliceRole:
```

File: pkg/prometheus/promcfg.go
Comment (ArthurSens) [inline pkg/prometheus/promcfg.go:102]:
```suggestion
	case monitoringv1.EndpointsRole:
```

File: pkg/prometheus/promcfg.go
Comment (ArthurSens) [inline pkg/prometheus/promcfg.go:71]: Do I understand correctly that you're removing the argument that tells if the k8s cluster where the operator is running supports EndpointSlices or not? Shouldn't this information be kept and used?

E.g. what happens if the user sets the role to EndpointSlice, but the cluster doesn't support it?

File: pkg/prometheus/promcfg_test.go
Comment (ArthurSens) [inline pkg/prometheus/promcfg_test.go:828]:
```suggestion
		role            monitoringv1.ServiceDiscoveryRole
```

File: pkg/prometheus/promcfg.go
Comment (ArthurSens) [inline pkg/prometheus/promcfg.go:116]: What I'm imagining is that the argument `endpointSliceSupported` still exists, that tells if the cluster itself supports endpointSlice, and `endpointSliceConfigured` is what you're checking with the switch case above.

```suggestion
		useEndpointSlice:       endpointSliceSupported && endpointSliceConfigured,
```

**Review by ArthurSens — COMMENTED** [review]
Thanks a lot for the patience and determination in this PR :)

I've made a few comments, one because I've noticed we created an Enum for the ServiceDiscoveryRole, but we aren't using it anywhere.

The second comment I'm not really sure if I'm missing something, but I have the impression that we check if the cluster supports the endpointslice role, but it seems like we're not using it for anything. 🤔

**Comment by ArthurSens** [comment]
The tests shows that we're not checking for `nil` somewhere in `NewConfigGenerator` 😬

File: pkg/prometheus/promcfg.go
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:98]:
```suggestion
	switch serviceDiscoveryRole := ptr.Deref(cpf.ServiceDiscoveryRole, monitoringv1.Endpoints); serviceDiscoveryRole {
```

**Review by simonpasquier — COMMENTED** [review]
This is getting in the right direction! I concur with @ArthurSens comment: we still need to pass the information to NewConfigGenerator() about the API supporting the Endpointslice resource or not.

File: pkg/prometheus/promcfg.go
Comment (thaniri) [reply pkg/prometheus/promcfg.go]: Thank you for your suggestion. I have implemented this in [7c1adf2](https://github.com/prometheus-operator/prometheus-operator/pull/6672/commits/7c1adf2433cbba2fb6c1bd84d736450767c03110)

File: pkg/prometheus/promcfg.go
Comment (thaniri) [reply pkg/prometheus/promcfg.go]: Thank you for the code style suggestion. I have implemented this in [7d5c6a5](https://github.com/prometheus-operator/prometheus-operator/pull/6672/commits/7d5c6a5421819696a553ae02c0af9df0f6b2beb2)

File: pkg/prometheus/promcfg.go
Comment (thaniri) [reply pkg/prometheus/promcfg.go]: Thank you for the code style suggestion. I have implemented this in [7d5c6a5](https://github.com/prometheus-operator/prometheus-operator/pull/6672/commits/7d5c6a5421819696a553ae02c0af9df0f6b2beb2)

File: pkg/prometheus/promcfg.go
Comment (thaniri) [reply pkg/prometheus/promcfg.go]: Thank you for pointing this out. It was an oversight on my part. Your comment [here](https://github.com/prometheus-operator/prometheus-operator/pull/6672#discussion_r1662943437) helped me realize the mistake I made.

I have re-added the argument to pass through the support of endpointslices here [16dc1f3](https://github.com/prometheus-operator/prometheus-operator/pull/6672/commits/16dc1f3d8a62d9076512e7e47993d7153177c4a9)

File: pkg/prometheus/promcfg_test.go
Comment (thaniri) [reply pkg/prometheus/promcfg_test.go]: Thank you for pointing this out. I have fixed this in [16dc1f3](https://github.com/prometheus-operator/prometheus-operator/pull/6672/commits/16dc1f3d8a62d9076512e7e47993d7153177c4a9)

File: pkg/prometheus/promcfg.go
Comment (thaniri) [reply pkg/prometheus/promcfg.go]: This is a wonderful comment that points out the missed logic in my implementation. I have fixed this in [16dc1f3](https://github.com/prometheus-operator/prometheus-operator/pull/6672/commits/16dc1f3d8a62d9076512e7e47993d7153177c4a9)

**Comment by thaniri** [comment]
> The tests show that we're not checking for `nil` somewhere in `NewConfigGenerator` 😬

I have fixed this now. Thank you to the reviewers for the patience in implementing this PR. It's summer where I live and as such holiday breaks are being taken :D

Tagging @simonpasquier and @ArthurSens directly to hopefully get a new review in!

File: pkg/prometheus/agent/operator.go
Comment (ArthurSens) [inline pkg/prometheus/agent/operator.go:325]: I guess we could avoid a variable assignment here. Do we really need the extra variable `k8sAPIEndpointSliceSupported`?

```go
	o.endpointSliceSupported, err := k8sutil.IsAPIGroupVersionResourceSupported(o.kclient.Discovery(), schema.GroupVersion{Group: "discovery.k8s.io", Version: "v1"}, "endpointslices")
	if err != nil {
		level.Warn(o.logger).Log("msg", "failed to check if the API supports the endpointslice resources", "err ", err)
	}
	level.Info(o.logger).Log("msg", "Kubernetes API capabilities", "endpointslices", o.endpointSliceSupported)
```

File: pkg/prometheus/promcfg.go
Comment (ArthurSens) [inline pkg/prometheus/promcfg.go:108]:
```suggestion
```
This is no-op because the value is already false :)

File: pkg/prometheus/promcfg.go
Comment (ArthurSens) [inline pkg/prometheus/promcfg.go:113]:
```suggestion
```

Same thing here

File: pkg/prometheus/server/operator.go
Comment (ArthurSens) [inline pkg/prometheus/server/operator.go:362]: Same comments here as I've made above :)

**Review by ArthurSens — COMMENTED** [review]
I didn't have much time to go too deep into this review, so please expect another one in a few days. Simon is also on PTO this week, I'd prefer to wait for him to comeback before merging anything anyway 😬

File: pkg/prometheus/agent/operator.go
Comment (thaniri) [reply pkg/prometheus/agent/operator.go:325]: I have implemented this suggestion. It reduces the LoC by 1 and saves 1 byte of memory each time `sync()` gets called.

File: pkg/prometheus/promcfg.go
Comment (thaniri) [reply pkg/prometheus/promcfg.go:108]: Line 99 is also a no-op as we assign false to the bool `endpointSliceConfigured`.

Perhaps we keep this no-op because I believe being explicit about intention instead of implicit is more helpful to future readers of the code.

File: pkg/prometheus/promcfg.go
Comment (thaniri) [reply pkg/prometheus/promcfg.go:113]: See comment https://github.com/prometheus-operator/prometheus-operator/pull/6672#discussion_r1681836580

File: pkg/prometheus/server/operator.go
Comment (thaniri) [reply pkg/prometheus/server/operator.go:362]: I have implemented this as in https://github.com/prometheus-operator/prometheus-operator/pull/6672#discussion_r1681835729

File: pkg/prometheus/promcfg.go
Comment (ArthurSens) [reply pkg/prometheus/promcfg.go:108]: Fair point :)

**Review by ArthurSens — DISMISSED** [review]
LGTM, thanks for your patience! I'm hoping Simon and Ha Ahn also express their opinion before merging here :)

**Comment by thaniri** [comment]
> LGTM, thanks for your patience! I'm hoping Simon and Ha Ahn also express their opinion before merging here :)

Thank you for your attentiveness and review @ArthurSens. I will wait for at least one more pull request approval. Please let me know if it is appropriate to directly message Simon or Ha Ahn.

**Review by simonpasquier — APPROVED** [review]
Thanks for the hard work! Tested locally and it works but one needs to grant the permissions on the endpointslices resource to the prometheus SA. I'll submit a follow-up PR to adjust the default manifests. I've got a few stylistic remarks but they don't warrant to delay merging this PR. The last thing would be to resolve the conflict.

**Comment by thaniri** [comment]
Thanks for helping with getting the code to a shippable state.

I've resolved the merge conflict, could one of you, @simonpasquier or @ArthurSens, please approve again? (And merge if you'd like because we are in different timezones)

**Review by simonpasquier — APPROVED** [review]
Thanks!

**Pull request merged**
